source/MRMesh/MRColorMapAggregator.h

```diff
@@ -0,0 +1,160 @@
+#pragma once
+#include "MRMeshFwd.h"
+#include "MRVector.h"
+#include "MRColor.h"
+#include "MRId.h"
+#include "MRBitSet.h"
+
+
+namespace MR
+{
+
+template<typename Tag>
+class ColorMapAggregator
+{
+public:
+    using ColorMap = Vector<Color, Id<Tag>>;
+    using ElementsBitSet = TaggedBitSet<Tag>;
+
+    ColorMapAggregator() = default;
+
+    void setDefaultColor( const Color& color );
+
+    void setColorMap( int i, const ColorMap& colorMap, const ElementsBitSet& elementsBitSet );
+
+    void resetColorMap( int i );
+    
+    enum class AggregateMode
+    {
+        Overlay,
+        Blending
+    };
+
+    void setMode( AggregateMode mode );
+
+    ColorMap aggregate();
+private:
+    Color defaultColor_;
+    static const int MaxColorMap{ 10 };
+
+    struct Data
+    {
+        bool active = false;
+        ColorMap colorMap;
+        ElementsBitSet elements;
+    };
+    std::vector<Data> dataSet_{ MaxColorMap };
+
+    ColorMap aggregatedColorMap_;
+    int colorMapSize_{ 0 };
+    bool needUpdate_{ true };
+    AggregateMode mode_{ AggregateMode::Overlay };
+
+    void updateAggregated();
+};
+
+template<typename Tag>
+void ColorMapAggregator<Tag>::setDefaultColor( const Color& color )
+{
+    defaultColor_ = color;
+}
+
+template<typename Tag>
+void ColorMapAggregator<Tag>::setColorMap( int i, const ColorMap& colorMap, const ElementsBitSet& elementsBitSet )
+{
+    assert( i >= 0 && i < MaxColorMap );
+    assert( !colorMap.empty() );
+    assert( colorMap.size() == elementsBitSet.size() );
+    if ( colorMapSize_ == 0 )
+        colorMapSize_ = int( colorMap.size() );
+    else
+        assert( colorMap.size() == colorMapSize_ );
+
+    dataSet_[i] = {true, colorMap, elementsBitSet};
+}
+
+template<typename Tag>
+void ColorMapAggregator<Tag>::resetColorMap( int i )
+{
+    assert( i >= 0 && i < MaxColorMap );
+    dataSet_[i].active = false;
+    needUpdate_ = true;
+
+    if ( !std::any_of( dataSet_.begin(), dataSet_.end(), [] ( auto e )
+    {
+        return e.active;
+    } ) )
+        colorMapSize_ = 0;
+}
+
+template<typename Tag>
+void ColorMapAggregator<Tag>::setMode( AggregateMode mode )
+{
+    if ( mode == mode_ )
+        return;
+    mode_ = mode;
+    needUpdate_ = true;
+}
+
+template<typename Tag>
+typename ColorMapAggregator<Tag>::ColorMap ColorMapAggregator<Tag>::aggregate()
+{
+    if ( needUpdate_ )
+        updateAggregated();
+    return aggregatedColorMap_;
+}
+
+template<typename Tag>
+void ColorMapAggregator<Tag>::updateAggregated()
+{
+    aggregatedColorMap_.clear();
+    aggregatedColorMap_.resize( colorMapSize_, defaultColor_ );
```

Comment:
pls try to hide implemetations in cpp file (you can find example in MRAABBTreeMaker.h and cpp )

Reply:
fixed